addon/addon/models/animation-participant.ts

```diff
@@ -32,39 +32,129 @@ export class AnimationParticipantManager {
   // In which case we don't want to call it during updateParticipants but instead
   // have the animations service call this
   performCleanup() {
-    for (let animationParticipant of this.participants) {
-      // Naive cleanup for now.
-      // This should be changed to something that can prune DOMRefs AND graft live nodes
-      // It should collect all DOMRefs to dispose, then traverse the tree and remove them at one go, preserving
-      // nodes that are NOT disposed and all of their descendants
-      animationParticipant._DOMRefsToDispose.forEach((DOMRef) => {
-        DOMRef.delete();
-      });
+    let unanimatedDetachedDOMRefs = new Set<DOMRefNode>();
+    let displacedDOMRefs = new Set<DOMRefNode>();
+    for (let participant of this.participants) {
+      participant._DOMRefsToDispose.forEach((node) =>
+        displacedDOMRefs.add(node)
+      );
+      participant._DOMRefsToDispose.clear();
+      if (
+        participant.uiState.detached &&
+        (!participant.uiState.detached.animation ||
+          participant.uiState.detached.animation.playState === 'finished' ||
+          participant.uiState.detached.animation?.playState === 'idle' ||
+          participant.uiState.detached.animation?.playState === 'paused')
+      ) {
+        // TODO: when cloning is implemented, add the DOMRef here
+        unanimatedDetachedDOMRefs.add(participant.uiState.detached.DOMRef);
+      }
+    }
```

Comment:
Record detached DOMRefNodes that are currently not used for any animations, and DOMRefNodes that are disposable because they are detached and another detached element has pushed them out.

We don't think about the hierarchy and how it affects them yet, we'll do that in a separate section.